src/storage/rdb.cc

```diff
@@ -449,3 +510,188 @@ Status RDB::Restore(const std::string &key, uint64_t ttl_ms) {
+      GET_OR_RETWITHLOG(loadObjectLen(nullptr));  // expires_size
+      continue;
+    } else if (type == RDBOpcodeAux) {
+      /* AUX: generic string-string fields. Use to add state to RDB
+       * which is backward compatible. Implementations of RDB loading
+       * are required to skip AUX fields they don't understand.
+       *
+       * An AUX field is composed of two strings: key and value. */
+      auto key = GET_OR_RETWITHLOG(LoadStringObject());
+      auto value = GET_OR_RETWITHLOG(LoadStringObject());
+      continue;
+    } else if (type == RDBOpcodeModuleAux) {
+      LOG(WARNING) << "RDB module not supported";
+      return {Status::NotOK, "RDB module not supported"};
+    } else if (type == RDBOpcodeFunction || type == RDBOpcodeFunction2) {
+      LOG(WARNING) << "RDB function not supported";
+      return {Status::NotOK, "RDB function not supported"};
+    } else {
+      if (!isObjectType(type)) {
+        LOG(WARNING) << "Invalid or Not supported object type: " << type;
+        return {Status::NotOK, "Invalid or Not supported object type"};
+      }
+    }
+
+    auto key = GET_OR_RETWITHLOG(LoadStringObject());
+    auto value = GET_OR_RETWITHLOG(loadRdbObject(type, key));
+
+    if (db_index != db_id) {  // skip db not match
+      continue;
+    }
+
+    if (isEmptyRedisObject(value)) {  // compatible with empty value
+      /* Since we used to have bug that could lead to empty keys
+       * (See #8453), we rather not fail when empty key is encountered
+       * in an RDB file, instead we will silently discard it and
+       * continue loading. */
+      if (empty_keys_skipped++ < 10) {
+        LOG(WARNING) << "skipping empty key: " << key;
+      }
+      continue;
+    } else if (expire_time != 0 &&
+               expire_time < now) {  // in redis this used to feed this deletion to any connected replicas
+      expire_keys++;
+      continue;
+    }
+
+    if (is_nx) {  // only load not exist key
+      auto s = exist(key);
+      if (!s.IsNotFound()) {
+        skip_exist_keys++;  // skip it even it's not okay
+        if (!s.ok()) {
+          LOG(ERROR) << "check key " << key << " exist failed: " << s.ToString();
+        }
+        continue;
+      }
+    }
+
+    auto ret = saveRdbObject(type, key, value, expire_time);
+    if (!ret.IsOK()) {
+      LOG(WARNING) << "save rdb object key " << key << " failed: " << ret.Msg();
+    } else {
+      load_keys++;
+    }
+  }
+
+  // Verify the checksum if RDB version is >= 5
+  if (rdb_ver >= 5) {
+    uint64_t chk_sum = 0;
+    auto expected = GET_OR_RETWITHLOG(stream_->GetCheckSum());
+    GET_OR_RETWITHLOG(stream_->Read(reinterpret_cast<char *>(&chk_sum), 8));
+    if (chk_sum == 0) {
+      LOG(WARNING) << "RDB file was saved with checksum disabled: no check performed.";
+    } else if (chk_sum != expected) {
+      LOG(WARNING) << "Wrong RDB checksum expected: " << chk_sum << " got: " << expected;
+      return {Status::NotOK, "Wrong RDB checksum"};
```

Comment:
So we check the checksum to return an error? What would be the result? As I understand from the code, all objects will be loaded and the function `RDB::LoadRdb` will return an error without any rollback. Can we write something like "All objects were processed and loaded but the checksum is unexpected" in the response to the client?

Reply:
Yes, if the checksum does not match, we will not perform a rollback. In Redis, if a checksum error occurs, the server will exit during start-up. However, as a command to load the RDB, it is not suitable to exit. Instead, I will add an error message to be sent to the client.